File: ch05/34/GetUglyNumber_Solution.cpp
class Solution {
public:
    int min(int data2, int data3, int data5){
    	int m = data2;
        if(data3 < m)
            m = data3;
        if(data5 < m)
            m = data5;
        return m;
    }
    
    int GetUglyNumber_Solution(int index) {
        if(index <=0)
            return 0;
        int uglyNum[index];
        uglyNum[0] = 1;
        int t2 = 0, t3 = 0, t5 = 0, data2, data3, data5;
        for(int i = 1; i<index; i++){
            int j2, j3, j5;
            for(j2 = t2; j2<i; j2++){
                data2 = uglyNum[j2]*2;
                if(data2 > uglyNum[i-1])
                    break;
            }
            
            for(j3 = t3; j3<i; j3++){
                data3 = uglyNum[j3]*3;
                if(data3 > uglyNum[i-1])
                    break;
            }
            
            for(j5 = t5; j5<i; j5++){
                data5 = uglyNum[j5]*5;
                if(data5 > uglyNum[i-1])
                    break;
            }
            
            int minOf235 = min(data2, data3, data5);
            if(minOf235 == data2)
            	t2 = j2;
            if(minOf235 == data3)
            	t3 = j3;
            if(minOf235 == data5)
            	t5 = j5;
            
            uglyNum[i] = minOf235;
        }
        return uglyNum[index - 1];
    }
};

File: ch05/34/GetUglyNumber_Solution.py
# -*- coding:utf-8 -*-
#####题目######
# 把只包含因子2、3和5的数称作丑数（Ugly Number）。
# 例如6、8都是丑数，但14不是，因为它包含因子7。 习惯上我们把1当做是第一个丑数。
# 求按从小到大的顺序的第N个丑数

#####解法#####
# m2代表乘以二刚好大于当前最大值的丑数的index
# m3代表乘以3刚好大于当前最大值的丑数的index

class Solution: 
    def GetUglyNumber_Solution(self, index):
        # write code here
        ugly_num = [1]
        if index <= 0:
            return 0
            
        ugly_len = 1
        m2 = m3 = m5 = 0
        while ugly_len < index:
            cur_max = ugly_num[-1]
            while ugly_num[m2] * 2 <= cur_max:
                m2 += 1
            while ugly_num[m3] * 3 <= cur_max:
                m3 += 1
            while ugly_num[m5] * 5 <= cur_max:
                m5 += 1

            min_of_m = min([ugly_num[m2] * 2, ugly_num[m3] * 3, ugly_num[m5] * 5])
            ugly_num.append(min_of_m)
            ugly_len += 1

        print ugly_num
        return ugly_num[index - 1]

if __name__ == "__main__":
    s = Solution()
    print s.GetUglyNumber_Solution(11)